File: subarrays_with_k_different_integers.cpp
class Solution {
public:
    int findAtMostK(int n, int k, vector<int>&nums){
        int ans = 0, left = 0, right = 0;
        unordered_map<int, int> mp;

        while(right < n){
            mp[nums[right]]++;

            while(mp.size() > k && left <= right){
                mp[nums[left]]--;
                if(mp[nums[left]] == 0) {
                    mp.erase(nums[left]);
                }
                left++;
            }
            ans += right-left+1;
            right++;
        }
        return ans;
    }

    int subarraysWithKDistinct(vector<int>& nums, int k) {
        int n = nums.size();
        return findAtMostK(n, k, nums) - findAtMostK(n, k-1, nums);
    }
};

// Time Complexity : O(n)
// Space Complexity : O(n)

File: subarrays_with_k_different_integers.go
func findAtMostK(n int, k int, nums []int) int {
    ans, left, right := 0, 0, 0
    mp := make(map[int]int)

    for right < n {
        mp[nums[right]]++

        for len(mp) > k && left <= right {
            mp[nums[left]]--

            if mp[nums[left]] == 0 {
                delete(mp, nums[left])
            }
            left++
        }
        ans += right-left+1
        right++
    }
    return ans
}

func subarraysWithKDistinct(nums []int, k int) int {
    n := len(nums)
    return findAtMostK(n, k, nums) - findAtMostK(n, k-1, nums)
}


// Time Complexity : O(n)
// Space Complexity : O(n)
